### PR TITLE
[jsonpath] List support

### PR DESCRIPTION
**Pull request by paulianttila**
Json path result can be list of Boolean, Number, String values

Signed-off-by: Pauli Anttila <pauli.anttila@gmail.com>

### CODE REVIEW

**Comment by paulianttila** [comment]
Sometimes there is a use case where jsonpath transformation should return list of values, especially when using transformation action in rules.

**Comment by davidgraeff** [comment]
Looks good, but can you add an example for this to the readme please, as the readme is our documentation of what is possible with this transformation.

**Comment by openhab-5iver** [comment]
This looks to be reverting ESH #[4951](https://github.com/eclipse/smarthome/pull/4951).

https://community.openhab.org/t/jsonpath-transformation-service-vs-jsonpath/40439

**Comment by davidgraeff** [comment]
But this time we have multiple tests to catch that particular issue and the lists are handled separately and not by the json library. But I'm not sure.

**Comment by martinvw** [comment]
@openhab-5iver thanks for mentioning and remembering :-)

@sjka @ThomDietrich wdyt?

File: bundles/org.openhab.transform.jsonpath/src/main/java/org/openhab/transform/jsonpath/internal/JSonPathTransformationService.java [hunk 1]
```diff
@@ -80,11 +81,23 @@ private String flattenList(List<?> list) {
...
         }
         if (list.size() > 1) {
+            if (list.get(0) instanceof Number || list.get(0) instanceof Boolean) {
+                return createNumberList(list);
+            } else if (list.get(0) instanceof String) {
+                return createStringList(list);
+            }
             logger.warn(
-                    "JsonPath expressions with more than one result are not allowed, please adapt your selector. Result: {}",
+                    "JsonPath expressions with more than one result is only supported for Boolean, Number and String data types, please adapt your selector. Result: {}",
```
Comment (J-N-K) [inline bundles/org.openhab.transform.jsonpath/src/main/java/org/openhab/transform/jsonpath/internal/JSonPathTransformationService.java]:
```suggestion
                    "JsonPath expressions with more than one result are only supported for Boolean, Number and String data types, please adapt your selector. Result: {}",
```

**Comment by J-N-K** [comment]
Where are we here? Since the change is covered by tests I would vote for merge.

**Comment by openhab-5iver** [comment]
Since @sjka and @ThomDietrich haven't replied, here are their arguments AGAINST...

https://github.com/eclipse/smarthome/issues/5132#issuecomment-369502756
https://github.com/eclipse/smarthome/issues/5132#issuecomment-369576824

IMO, JSON strings are a valid data type for Items, the neutered JSONPATH TS from 4951 is barely usable in rules since it errors when using indefinite paths that return more than one element in the list, and this PR should be merged. @Josar, @jboeddeker, @rkoshak?

**Comment by rkoshak** [comment]
I see both sides of the argument. But I can say based on fielding questions on the forum, users expect JSONPATH  to work like standard JSONPATH expressions. I can't say how many times I've had to deal with "it works on this jsonpath checker online, why doesn't it work with OH"?

But I also see how setting an Item's state to a json array. I can see how that would be considered a misuse of Items. But the alternative presented is to go ahead and let the Item become the full JSON in some cases and process the result in a Rule. So even the attempt to protect the user form himself requires, in some cases, misusing an Item.

Add to that bindings like MQTT2 allow chaining of transformations, I'm simply not convinced that we should alter the behavior of jsonpath to protect the users from themselves. It's better for the transform to work as expected per the standard than neutering it to protect the users from themselves.

**Comment by J-N-K** [comment]
@paulianttila, after reading that most agree with my opinion (should be merged), I would approve and merge after you applied my suggestion. Could you do that?

File: bundles/org.openhab.transform.jsonpath/src/main/java/org/openhab/transform/jsonpath/internal/JSonPathTransformationService.java (hunk 1 above)
Comment (paulianttila) [reply bundles/org.openhab.transform.jsonpath/src/main/java/org/openhab/transform/jsonpath/internal/JSonPathTransformationService.java]: Done

**Comment by Josar** [comment]
Just to raise awarness, it was not waited for the people whom advocated against a full json implementation to answer, imho. I was in favour of a properly working json path implementation so i welcome the changes.

Also, did someone think about the documentation? 
https://www.openhab.org/addons/transformations/jsonpath/

**Comment by J-N-K** [comment]
@Josar there were several weeks from raising to merging the PR. What timespan do you feel is appropriate?

**Comment by Josar** [comment]
Just my two cents, it would not have hurt to wait for the "con" advocates to at least answer and trigger them multiple times, as there might be a vacation, stress etc. that keeps them afk. I see that this was done, but still had a bad feeling as i only see the "pro" people taking part. 2 month, from PR start and 1 month from 1 triggering @sjka felt somehow very short. I have seen readme file changes beeing discussed more thouroughly than this PR. ;-)

**Comment by davidgraeff** [comment]
The project is moving too SLOW, not too fast :D 7 weeks was enough. If people had serious considerations they could have voiced them. Or do it now and we roll back if necessary.

**Comment by paulianttila** [comment]
@Josar, when do you have seen sjka? Simon was very active in eclipse smarthome side, but as Deutsche Telekom retreat from the project I have not seen Simon since either.

**Comment by Josar** [comment]
OK, if thats the case, i am not into details and just stated my observation.

**Pull request merged**